src/v/kafka/server/offset_translator.h

```diff
@@ -53,10 +53,30 @@ class offset_translator {
 private:
     int64_t delta(model::offset o) const {
         auto it = _cfg_mgr.lower_bound(o);
-        if (it != _cfg_mgr.begin()) {
-            --it;
+
+        if (it == _cfg_mgr.begin()) {
+            /**
+             * Special case for translation of start offset, after controll
+             * batches index is truncated first entry in configuration manager
+             * will have exactly the truncation offset, however this can not
+             * change the offset of batch that previously was placed at the
+             * truncation point.
```

Comment:
I still don't quite understand why this special case is needed. Looks like we will always have a configuration strictly before the start of the log (first at -infinity, then at the snapshot offset (which is start_of_log - 1)) so if we translate the log start offset we will never end up in this if branch.

Reply:
@mmaslankaprv what scenario exercises this branch?

Reply:
@dotnwat, @ztlpn, is right, it is exactly the case, I have some fuzzy tests for offset translation layer, they were failing since they are asking the translator layer about truncation offset, this is not the case during normal operation. I will adjust the tests accordingly